### ISSUE TITLE
abdapi ConnectionPool doesn't allow for synchronous close

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2680]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2680|
|Type|defect|
|Created|2007-05-28 08:03:21Z|
|Branch|https://github.com/twisted/twisted/tree/adbapi-close-before-start-2680|

In particular,

```python
def on_start():
    try:
        dbpool = ConnectionPool(...)
        dbpool.close()
    finally:
        reactor.stop()

reactor.callWhenRunning(on_start)
reactor.run()
```

fails with

```
  File ".../twisted/enterprise/adbapi.py", line 305, in close
    if self.shutdownID:
AttributeError: ConnectionPool instance has no attribute 'shutdownID'
```

Attachments:

* [adbapi-shutdownID-fix.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fa/1fa79bd9f47d73a47762a6c13c7e147ca8223855/bcf755162e7a2d138b0fb9e3d3915cca652caa9c.diff) (1489 bytes) - added by bdallen on 2009-04-19 22:39:00Z - svn diff for fix, including unit test
* [adbapi-shutdownID-fix-20090503.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1fa/1fa79bd9f47d73a47762a6c13c7e147ca8223855/402892573d0adf19af61691628ef5d1db2802ccc.diff) (5119 bytes) - added by bdallen on 2009-05-03 19:12:04Z - Updated patch which relocates shutdownID = None above callWhenRunning and test for removal of all triggers.

<details><summary>Searchable metadata</summary>

```
trac-id__2680 2680
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_adbapi_close_before_start_2680 branches/adbapi-close-before-start-2680
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__adbapi adbapi
time__1180339401000000 1180339401000000
changetime__1278690607000000 1278690607000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_1"></a>|@jdavisp3 set status to `closed`|
|-|-|

I think this has been fixed for a while. At least, the sample code works in Twisted 8.

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|bdallen set status to `reopened`|
|-|-|

This is still a problem if close() is called before start(). For example:

from twisted.enterprise import adbapi
pool = adbapi.ConnectionPool(...)
pool.close()

I think the reason this has remained unfixed is that calling close() before start() is abnormal usage. Still it is a bug - __init__ adds a startup trigger, and it should be possible to remove it by closeing even if start was not called. It's also trivial to fix by adding "self.shutdownID = None" to __init__. I will attach a diff.

On a related note, shouldn't ConnectionPool have a destructor which calls close?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Hmmm.  Why are you calling `close` before `start`?  That seems somewhat pointless to me.  What _should_ the behavior be in that case?

I'd be in favor of just documenting this case as invalid.

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|bdallen commented|
|-|-|

Replying to [exarkun](#note_3):
> Hmmm.  Why are you calling `close` before `start`?  That seems somewhat pointless to me.  What _should_ the behavior be in that case?
>
> I'd be in favor of just documenting this case as invalid.

I'm not calling close before start on purpose. I call close() in a database access layer class destructor, and an application using my db layer raised an uncaught exception before reactor.run() was called. The destructor gets called before python exits, and the shutdownID exception pops up. This is confusing, because there is really nothing wrong with the db access layer, the real error is the original exception.

This actually happened in real code, so I think it's worth fixing. It's also "the right thing", because there is nothing conceptually wrong with calling close() on a pool without a pool.start() or reactor.run().

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Okay.  I'm certainly in favor of fixing bugs in real applications. :)  I'm not sure I agree that it is "conceptually" sensible to close something before starting it (perhaps if these names actually lined up with each other - open and close, or start and stop - it would be easier for me to make that case, though).

In any case, I think I can see what should happen if close is called without start or run - nothing.  What should happen if close is called and then start is called, though?  Does the pool end up running or not running?

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|bdallen commented|
|-|-|

I guess I'm really talking about "close without start". Calling start after close really does smell like an error, although in the case of ConnectionPool it might actually work (assuming you can re-start() a ThreadPool after stop()ing). I'll do some testing later.

I think "close without start" should still remove the reactor start trigger, which is installed in __init__. Users would be very suprised if the reactor automatically started a ConnectionPool that they had explicitly close()ed. This is actually what happens with my one line patch. close already checks and only removes the shutdown and start event triggers if they evaluate to True. The problem is that self.shutdownID didn't even exist if start was never called.

Conceptually I think close methods should release all resources associated with an object, regardless of object state. Here that means closing all connection objects and reactor hooks and closing (i.e. stopping) the threadpool. If any of these resources have not been aquired, close should not try to remove them anyway and raise a cryptic exception. I think it shouldn't raise any exception because close without start is rare behavior suggesting an error in the calling application, possibly unrelated to ConnectionPool, not error behavior that breaks ConnectionPool.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Hmmm.  Right.  I forgot how `__init__` also plays along in this scenario.  So it's really a three phase process, not just a two phase start/close.  That's probably mainly why things are so weird.

So first you create a `ConnectionPool`.  At that point, it's scheduled to start when the reactor runs.  Then you close it, which raises an exception.  And if the reactor actually starts after that, then the `ConnectionPool` starts its threadpool.  When the reactor later stops, the `ConnectionPool` will also stop, since `start` added a trigger for that.

Alright, now that I'm caught up...

> I think it shouldn't raise any exception because close without start is rare behavior suggesting an error in the calling application, possibly unrelated to `ConnectionPool`, not error behavior that breaks `ConnectionPool`.

I don't agree with this justification.  A bug in application code which causes the mis-use of some library code is generally most easily tracked down when the library code makes this mis-use obvious - not when it quietly allows it to pass, trying to figure out something sensible do to even though it doesn't quite make sense to do anything.

Part of the problem here is that `ConnectionPool`'s API seems to lead to idiomatic code which results in bad things (currently, in `close` being called at the wrong time and raising an exception - even if we fix the exception, it still may be wrong: consider the case where `close` gets called, but the reactor starts up anyway; now you have a `ConnectionPool` which isn't running, and any attempt to use it is going to result in `Deferred`s which never fire.  This doesn't seem to be an improvement on the current behavior).

So I wonder if we can improve the API so that the idiomatic usage is less error prone.  First, to be clear, the current usage seems to be something like this:

```python
pool = ConnectionPool(...)
service = SomeService(pool)
randomOtherSetupCode()
reactor.run()
```

`SomeService` has cleanup code which closes the pool passed to it (apparently using a weakref or an `__del__` I suppose?) and `randomOtherSetupCode` raises an exception because it's buggy.  The result is that `SomeService` attempts to close the pool and gets an error thrown at it, possibly preventing it from running the rest of its cleanup code, and at least logging an unhandled exception which masks or at least confuses the real issue - that `randomOtherSetupCode` is buggy.

One thing about this I don't particularly like is the use of a weakref or `__del__` to clean up the ConnectionPool.  Is this really representative of the case you've got?  Actually, to be more specific, using `__del__` is probably a pretty bad idea (since it interferes with cycle collection), but I suppose using a weakref is reasonable.  I'd probably go with an API that requires the application to explicitly release the object (as is the case with `ConnectionPool` and many other classes in Twisted), but maybe that's just me.

Anyway, after all that rambling, I don't really have any good ideas for an improved API.  I still think the problem comes mostly from the split between `__init__` and `start`.  The trick is knowing whether the `ConnectionPool` is running or not.  I suppose this is possible by checking the `running` attribute.  However, the `close` method is still an attractive nuisance - it's easy to forget about (or never notice in the first place) `running` and just try to use close.

Then, from the usage pattern above, it seems like the right thing to do if `start` is called after `close` is for the `ConnectionPool` not to start.  The object which was going to make use of it has said that it doesn't want it anymore.  Normally, with the reactor trigger being responsible for calling `start`, this won't happen anyway, since `close` can remove the startup trigger.

So I like your patch.  Can you write a unit test covering this case to make sure it really behaves as we want?

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|bdallen commented|
|-|-|

Heh, I'm starting to feel guilty about using work time for this discussion. I'll create a unit test this weekend and think some more about the API. I do think it's worth patching this specific problem before making more disruptive API changes.

> I don't agree with this justification.  A bug in application code which causes the mis-use of some library code is generally most easily tracked down when the library code makes this mis-use obvious - not when it quietly allows it to pass, trying to figure out something sensible do to even though it doesn't quite make sense to do anything.

I agree with you there, but I think determining what is really a mis-use of the library code and where throwing an exception really helps the library user is tricky. For the close() without start() case, I think throwing an exception actually makes the library harder to use, and that case is not really a "mis-use". The real mis-use comes when they try to start() after close(), where throwing an exception would probably be helpful. I'll try to find some examples from other APIs and think about it a bit more.

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|bdallen commented|
|-|-|

I created a unit test in twisted.test.test_adbapi in the class ConnectionPoolTestCase. I was a little unsure about style and what message to fail with, but the test definitely fails without my patch and passes with the patch. The new diff includes both the fix and the test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to bdallen|
|-|-|

Thanks. :)  A few points:

  1. The unit test should take care to clean up whatever global state it plays with.  In this case, it should get rid of the key it adds to `sys.modules`.  This is easily done with `self.addCleanup(sys.modules.pop, 'FakeDBDriver')` in this case.  However, another possibility would be to just drop the `_FakeDBDriver` code into a module in `twisted/test/` and not mess around with `sys.modules` directly. (On further note, `class _FakeDBDriver():` is a syntax error on most versions of Python supported by Twisted - declare classes with the old syntax in contributions to Twisted - `class _FakeDBDriver:`.)
  1. The behavior of calling `start` after `close` is untested.  If you agree that this should raise an exception, it'd be good to test and implement that.
  1. I think the position of the new `self.shutdownID = None` statement isn't quite right.  If the `ConnectionPool` is created when the reactor is already running, the `_start` method will be called by `reactor.callWhenRunning` and _then_ `shutdownID` will get set to `None` - not what's desired.  Can you add a test for creating a `ConnectionPool` when the reactor is already running (this might be easier if you parameterize the reactor it uses and then pass in a fake with each of the desired behaviors).

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|bdallen commented|
|-|-|

Regarding #3, do you mean modifying ConnectionPool so it can use a custom reactor instead of twisted.internet.reactor? Otherwise I can't figure out how to create the test without interfering with the test runner - I had trouble when I tried temporarily replacing twisted.internet.reactor, and using the main reactor and calling stop seems to be frowned upon.

Right now _start does not touch shutdownID, so the placement at the end of __init__ works even if the reactor is running, but I agree that moving it up would be more robust to future changes.

**Comment by twisted-trac** [comment]
|<img alt="terrycojones's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|terrycojones commented|
|-|-|

I just ran into this too, in the context of writing unit tests.

I started with the simplest possible code, a setUp method that creates a connection pool, a tearDown method that calls close, and a test case that just calls pass. Running the test with trial, I get the same `ConnectionPool instance has no attribute 'shutdownID'` exception.

Looking at the code it seems adding `self.shutdownID = None` to `__init__` or using `hasattr(self, 'shutdownID')` in `close` would help. But I see involved discussion of that above, and I've not read it closely, so I'm sure I don't know the subtleties here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Replying to [bdallen](#note_12):
> Regarding #3, do you mean modifying ConnectionPool so it can use a custom reactor instead of twisted.internet.reactor? Otherwise I can't figure out how to create the test without interfering with the test runner - I had trouble when I tried temporarily replacing twisted.internet.reactor, and using the main reactor and calling stop seems to be frowned upon.

Yes, parameterizing the reactor to `ConnectionPool` is probably the best approach.

>
> Right now _start does not touch shutdownID, so the placement at the end of __init__ works even if the reactor is running, but I agree that moving it up would be more robust to future changes.

`_start` calls `start` and `start` sets `shutdownID` - so I would say that it does touch it, and the current placement of the `self.shutdownID = None` line at the end of `__init__` will result in the shutdown id being lost sometimes.

**Comment by twisted-trac** [comment]
|<img alt="bdallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|bdallen commented|
|-|-|

> `_start` calls `start` and `start` sets `shutdownID` - so I would say that it does touch it, and the current placement of the `self.shutdownID = None` line at the end of `__init__` will result in the shutdown id being lost sometimes.
Good point - removal of all event triggers is now tested in one of the unit tests, and fails if `self.shutdownID = None` is placed below callWhenRunning.

I did some testing, and running queries on a re-`start`ed (calling `start`, `close`, `start`) pool works. Now that I think about it, it may be useful - it allows you to temporarely close all the connections to release resources, and connect again later, i.e. use `close` like a `stop` function. On the other hand running queries on a `close`ed pool silently ignores all requests - this comes down to ThreadPool.callInThreadWithCallback, which does a `return` if `self.joined` is set, which is done in stop. This behavior may also be useful, at least for thread pool - you can use it to temporarily disable all calls that go through the thread pool. For ConnectionPool you might want to do the same thing - temporarily disable database access. On the other hand, we could raise an exception, and if they really want this behavior they can try..except: pass.

Right now I'm thinking we should throw an exception if run* is called on a `stop`ed pool, and allow the other cases without raising. If run* is called on a stopped pool, silently dropping the query is likely to break an application. For the other cases - `close`ing a not-running pool and re-`start`ing a closed pool - I can't think of any scenarios where they would actually break an application, and I can think of several where a developer would want to do them on purpose (for the first it would be done in a clean-up routine if the application executes pre-maturely). In other words I don't think they are errors at all, so it would be obnoxious to raise.

I think I've addressed the other issues in the latest patch; let me know what you think about the above discussion and I'll implement whatever solution we decide on.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph removed owner|
|-|-|

Reassigning to nobody to facilitate faster reviews.

(Reviews should be unassigned if you don't need someone in particular to review it.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to bdallen|
|-|-|

Thanks again for your contribution, and for sticking with the process thus far.  I apologize for the slow pace of reviews; the Twisted dev team has been stretched really thin lately.  We appreciate your patience.  It's getting close - stick with us just a bit longer and we should get it integrated :).

  1. Regarding the `test_reactor` argument:
    1. Please use the `cp_` prefix for it; the reason all the other parameters on `ConnectionPool` use that prefix is to avoid conflicting with parameters that are potentailly relevant to db-api drivers.
    1. It should just be `cp_reactor`.  If you have some other, non-testing-related reason to want to plug in a different `IReactorCore` provider, it should work.  In general, APIs should just say what they do, and do it correctly, whether the main area of interest is for testing or not.  For example, SQLite has in-memory databases, which are perfectly usable in production, even if most of the time when you want an in-memory DB it's just for testing.
    1. It should have a `@type` explaining that it is expected to be an `IReactorCore` provider.
    1. There should also be separate `@ivar` documentation on the class explaining that it has a `reactor` attribute.  In general we can skip the `__init__` docstring if the name of the `__init__` parameter and the instance variable match up, but in this case they don't and I think it would be good to be super explicit.
  1. The parentheses around `if ('test_reactor' in connkw)` and `if (fakeReactor.hasUnremovedEventTriggers())` are extraneous; not a violation of the coding standard, but perhaps we should make it one ;).
  1. `shutdownID` should also have an `@ivar` describing its purpose, as it is becoming better documented and supported in this change.
  1. There are failing tests:
```
===============================================================================
[ERROR]: twisted.test.test_adbapi.ConnectionPoolTestCase.test_closeWithoutStartAttributeErrorFix

Traceback (most recent call last):
  File "twisted/test/test_adbapi.py", line 781, in test_closeWithoutStartAttributeErrorFix
    pool = ConnectionPool('twisted.test.fakedb')
  File "twisted/enterprise/adbapi.py", line 190, in __init__
    self.dbapi = reflect.namedModule(dbapiName)
  File "twisted/python/reflect.py", line 337, in namedModule
    topLevel = __import__(name)
exceptions.ImportError: No module named fakedb
===============================================================================
[ERROR]: twisted.test.test_adbapi.ConnectionPoolTestCase.test_instantiateWhileReactorIsRunning

Traceback (most recent call last):
  File "twisted/test/test_adbapi.py", line 793, in test_instantiateWhileReactorIsRunning
    from twisted.test.fake_reactor import FakeReactor
exceptions.ImportError: No module named fake_reactor
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

[#4039](https://github.com/twisted/twisted/issues/4039) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [28210]) Branching to 'adbapi-close-before-start-2680'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner|
|-|-|

Addressed the feedback.

I ended up dropping the part of the patch which makes `ConnectionPool` use its `_reactor` attribute for the `deferToThreadPool` calls.  This is a nice change, but the unit tests don't cover it at all.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. Can you open a bug so that self._reactor is used instead of the global reactor for runInteraction/runWithConnection. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

Sorry for the delay.  Filed [#4559](https://github.com/twisted/twisted/issues/4559) and [#4560](https://github.com/twisted/twisted/issues/4560).  Merging now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `closed`|
|-|-|

(In [29734]) Merge adbapi-close-before-start-2680

Author: bdallen, exarkun
Reviewer: exarkun, glyph, therve
Fixes: [#2680](https://github.com/twisted/twisted/issues/2680)

Allow `ConnectionPool.close` to be called before `ConnectionPool.start`
to abort the scheduled startup.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
